### DISCUSSION TITLE
MarkerEnd does not work when passed conditionally to BaseEdge Component.

### DISCUSSION THREAD

**Discussion by surya0180** [xyflow/xyflow#3673] [Q&A]
I have a custom edge like this

````
import React from "react";
import {
  getBezierPath,
  EdgeLabelRenderer,
  BaseEdge,
  MarkerType,
} from "reactflow";

function ScenarioEdge({
  id,
  sourceX,
  sourceY,
  targetX,
  targetY,
  sourcePosition,
  targetPosition,
  data,
  markerEnd,
  style,
}) {
  const [edgePath, labelX, labelY] = getBezierPath({
    sourceX,
    sourceY,
    sourcePosition,
    targetX,
    targetY,
    targetPosition,
  });

  const canBeRendered = findCondition(); /* Some logic to determine if I want to have a color or have a default disabled color */

  if (!canBeRendered) {
    style = {
      strokeWidth: 2,
      stroke: "#9e9e9e",
    };
  }

  return (
    <>
      <BaseEdge
        id={id}
        path={edgePath}
        markerEnd={
          canBeRendered
            ? markerEnd
            : "url(#1__color=#9e9e9eff&height=20&type=arrow&width=20)"
        }
        style={{ ...style }}
      />
      {canBeRendered && (
        <EdgeLabelRenderer>{/* Some jsx goes here */}</EdgeLabelRenderer>
      )}
    </>
  );
}

export default ScenarioEdge; 
````

I am passing markerEnd conditionally as you can see in the code above. The style is working as expected, but the markerEnd is not even getting rendered on the screen. What might be the issue?

**Comment by moklick** [comment]
does it work without the condition?

**Reply by surya0180** [reply]
Hi @moklick, Sorry for the late reply as I was busy with work.

Yeah, it works without the condition. But when I add the condition, it doesn't work.

**Reply by bcakmakoglu** [reply]
Where did you get this from `url(#1__color=#9e9e9eff&height=20&type=arrow&width=20)`?

I'm assuming you want to use a different sort of marker than the one you get in `markerEnd`?
But just using a different URL doesn't mean that this marker actually exists - if no marker matches this URL, you won't see a marker; I'm thinking that's what Moritz meant when he asked if it works without the condition.

The marker urls are constructed by parsing the options of a marker in an edge (`edge.markerStart` / `edge.markerEnd`), so if no marker in any of your edges matches the properties that of the url you're trying to use, you won't see a marker, since there is no marker that would be linked to that url 😄 

In this [sandbox](https://codesandbox.io/p/sandbox/black-wind-jyn5wg?file=%2FButtonEdge.tsx%3A36%2C20-36%2C63) for example you can see that you can switch between different markers, *given that the marker definition actually exists*.

So I have these two edges 

```ts
{
    id: "edge-button",
    source: "button-1",
    target: "button-2",
    type: "buttonedge",
    markerEnd: {
      type: MarkerType.ArrowClosed,
      color: "pink",
    },
  },
  {
    id: "edge-button",
    source: "button-3",
    target: "button-4",
    type: "buttonedge",
    markerEnd: {
      type: MarkerType.ArrowClosed,
      color: "red",
    },
  },
  ```

that means I can switch between these to marker urls `url(#1__color=red&type=arrowclosed)` and `url(#1__color=pink&type=arrowclosed)`

Trying something like this `url(#1__color=purple&type=arrowclosed)` would not work, since I never created a marker definition with the color `purple` to begin with, so there won't be any marker to reference like this.

The URL isn't meant for you to modify so you can easily change colors or whatever, that's not how the marker urls work, these are references to existing markers matching exactly the definition that was used for that marker.

**Reply by surya0180** [reply]
In my use case:
1. Each node represents some data
2. If the data is not present completely, all the edges that are coming towards that node will be gray. I am able to change the color of the edge but I failed to change the color of the marker. 
3. According to what you said, If I were to change the color of the marker, That kind of marker should already be present in the edges. But in my case, It is not always guaranteed that such a marker exists.

Is there any way that I can use a color that is not present in any of the existing edge markers?

**Reply by bcakmakoglu** [reply]
I'd think it's only possible if you start pre-defining your markers yourself, even if no edge using it exists, by setting up your own marker-definitions outside of ReactFlow and referencing those - similar to what happens in the [`EdgeMarkers`](https://reactflow.dev/examples/edges/markers) example using a custom edge marker.

```ts
<svg style={{ position: 'absolute', top: 0, left: 0 }}>
	<defs>
		<marker
		      className="react-flow__arrowhead"
			  id="markerWithColor"
		      markerWidth={12.5}
		      markerHeight={12.5}
		      viewBox="-10 -10 20 20"
		      markerUnits="strokeWidth"
		      orient="auto-start-reverse"
		      refX="0"
		      refY="0"
		    >
				<polyline
					style={{
					  stroke: color,
					  strokeWidth,
					}}
					strokeLinecap="round"
					strokeLinejoin="round"
					fill="none"
					points="-5,-4 0,0 -5,4"
				/>

**Reply by surya0180** [reply]
This works for now. Thanks for the help @bcakmakoglu @moklick ❤️ 😄 .
